src/AspNetCoreModuleV2/CommonLib/FileOutputManager.cpp

```
@@ -0,0 +1,155 @@
+                // Read file fails.
+                if (ReadFile(m_hLogFileHandle, pzFileContents, 4096, &dwNumBytesRead, NULL))
+                {
+                    if (SUCCEEDED(struStdOutput->CopyA(pzFileContents, dwNumBytesRead)))
+                    {
+                        fLogged = TRUE;
+                    }
+                }
+            }
+        }
+    }
+
+    return fLogged;
+}
+
+FileOutputManager::FileOutputManager(PCWSTR pwzStdOutLogFileName, PCWSTR pwzApplicationPath)
+    : m_hLogFileHandle(INVALID_HANDLE_VALUE)
+{
+    m_wsApplicationPath = std::wstring(pwzApplicationPath);
+    m_wsStdOutLogFileName = std::wstring(pwzStdOutLogFileName);
+}
+
+HRESULT
+FileOutputManager::Start()
+{
+    HRESULT hr;
+    SYSTEMTIME systemTime;
+    SECURITY_ATTRIBUTES saAttr = { 0 };
+    STRU struPath;
+
+    FILE* pStdOutFile;
+
+    hr = UTILITY::ConvertPathToFullPath(
+        m_wsStdOutLogFileName.c_str(),
+        m_wsApplicationPath.c_str(),
+        &struPath);
+
+    if (FAILED(hr))
+    {
+        goto Finished;
+    }
+
+    hr = UTILITY::EnsureDirectoryPathExist(struPath.QueryStr());
+    if (FAILED(hr))
+    {
+        goto Finished;
+    }
+    m_fdStdOut = _dup(_fileno(stdout));
+    m_fdStdErr = _dup(_fileno(stderr));
+
+    GetSystemTime(&systemTime);
+    hr = m_struLogFilePath.SafeSnwprintf(L"%s_%d%02d%02d%02d%02d%02d_%d.log",
+        struPath.QueryStr(),
+        systemTime.wYear,
+        systemTime.wMonth,
+        systemTime.wDay,
+        systemTime.wHour,
+        systemTime.wMinute,
+        systemTime.wSecond,
+        GetCurrentProcessId());
+    if (FAILED(hr))
+    {
+        goto Finished;
+    }
+    saAttr.nLength = sizeof(SECURITY_ATTRIBUTES);
+    saAttr.bInheritHandle = TRUE;
+    saAttr.lpSecurityDescriptor = NULL;
+
+    if (_wfreopen_s(&pStdOutFile, m_struLogFilePath.QueryStr(), L"w+", stdout) == 0)
+    {
+        setvbuf(pStdOutFile, NULL, _IONBF, 0);
+        _dup2(_fileno(pStdOutFile), _fileno(stdout));
```

you may want to check whether _dup2 success or not

---

If dup2 fails, I don't see what else we could do.

---

you can at least fail faster and skip some stack